pytorch_lightning/trainer/data_loading.py

```diff
@@ -285,7 +285,11 @@ def _reset_eval_dataloader(
         # datasets could be none, 1 or 2+
         if len(dataloaders) != 0:
             for i, dataloader in enumerate(dataloaders):
-                num_batches = 0
+                try:
+                    num_batches = len(dataloader)
+                except (TypeError, NotImplementedError):
+                    num_batches = float('inf')
```

Comment:
```suggestion
                num_batches = len(dataloader) if callable(getattr(dataloader, "__len__")) else float("inf")
```
Wouldn't that be better?

Reply:
I think it is bad to use exceptions for control flow like this. If the user implements `len` but has a type error, we catch it and the program crashes somewhere else, making it much harder for the user to debug.

Reply:
seems like the suggestion did not work, so reverting it back... :(

Reply:
:(

Reply:
what was the error?

Reply:
what about `num_batches = len(dataloader) if _has_len(dataloader) else float('inf')`?

Reply:
This seems to be a longterm solution.

Reply:
@awaelchli see here - https://github.com/PyTorchLightning/pytorch-lightning/runs/812000184

Reply:
the problem is raising an exception in the `__len__`

Reply:
yeah my orignal suggestions was wrong sorry. my new suggestion is to use the already existing _has_len method
`num_batches = len(dataloader) if _has_len(dataloader) else float('inf')`
but not a big deal, would just be cleaner.